kratos/processes/compute_nodal_gradient_process.cpp

```diff
@@ -269,20 +292,30 @@ ComputeNodalGradientProcess<ComputeNodalGradientProcessSettings::SaveAsNonHistor
 {
     KRATOS_TRY
 
-    // Doing several checks
+    auto& r_nodes = rModelPart.Nodes();
+
+    // Additional checks
     if (!mNonHistoricalVariable) {
-        VariableUtils().CheckVariableExists(rOriginVariable, mrModelPart.Nodes());
+        // Not sure about this, seems to be doing nothing.
+        VariableUtils().CheckVariableExists(rOriginVariable, r_nodes);
     } else {
-        KRATOS_ERROR_IF_NOT(mrModelPart.Nodes().begin()->Has(rOriginVariable)) << "Variable " << rOriginVariable.Name() << " not defined on non-historial database" << std::endl;
+        bool hasVariable = 0;
+        if(mrModelPart.NumberOfNodes() != 0)
+            hasVariable = mrModelPart.Nodes().begin()->Has(rOriginVariable);
+        hasVariable = mrModelPart.GetCommunicator().GetDataCommunicator().MaxAll(hasVariable);
+
+        KRATOS_ERROR_IF_NOT(hasVariable) << "Variable " << rOriginVariable.Name() << " not defined on non-historial database" << std::endl;
     }
+
     // In case the area or gradient variable is not initialized we initialize it
-    auto& r_nodes = rModelPart.Nodes();
-    if (!r_nodes.begin()->Has( rGradientVariable )) {
-        const array_1d<double,3> zero_vector = ZeroVector(3);
-        VariableUtils().SetNonHistoricalVariable(rGradientVariable, zero_vector, r_nodes);
-    }
-    if (!r_nodes.begin()->Has( rAreaVariable )) {
-        VariableUtils().SetNonHistoricalVariable(rAreaVariable, 0.0, r_nodes);
+    if (mrModelPart.NumberOfNodes() != 0) {
+        if (!r_nodes.begin()->Has( rGradientVariable )) {
+            const array_1d<double,3> zero_vector = ZeroVector(3);
+            VariableUtils().SetNonHistoricalVariable(rGradientVariable, zero_vector, r_nodes);
+        }
+        if (!r_nodes.begin()->Has( rAreaVariable )) {
+            VariableUtils().SetNonHistoricalVariable(rAreaVariable, 0.0, r_nodes);
+        }
```

Comment:
the code is copy pasted from above no?
=> can you refactor to reduce duplication?

Reply:
True, done!

I left the Parameters() constructors untouched as they were not working and require new tests, which is covered in #7569